secrets/certregistry/certregistry.go

```diff
@@ -0,0 +1,218 @@
+package certregistry
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+	"strings"
+	"sync"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+)
+
+var (
+	defaultHost = "ingress.local"
+	currentTime = time.Now
+	// ErrNoMatchingCertificateFound is used if there is no matching certificate found
+	ErrNoMatchingCertificateFound = errors.New("no matching certificate found")
+)
+
+const (
+	// used as wildcard char in Cert Hostname/AltName matches
+	glob = "*"
+	// minimal time period for the NotAfter attribute of a Cert to be in the future
+	minimalCertValidityPeriod = 7 * 24 * time.Hour
+)
+
+type tlsCertificate struct {
+	hosts []string
+	cert  *tls.Certificate
+}
+
+// CertRegistry object holds TLS certificates to be used to terminate TLS connections
+// We ensure ensure syncronized access to them and hold a default certificate.
+type CertRegistry struct {
+	lookup         map[string]*tlsCertificate
+	mx             *sync.Mutex
+	defaultTLSCert *tls.Certificate
+}
+
+// NewCertRegistry initializes the certificate registry with an empty map
+// and a generated default certificate.
+func NewCertRegistry() *CertRegistry {
+	l := make(map[string]*tlsCertificate)
+
+	return &CertRegistry{
+		lookup:         l,
+		mx:             &sync.Mutex{},
+		defaultTLSCert: getFakeHostTLSCert(defaultHost),
+	}
+}
+
+func (r *CertRegistry) getCertByKey(key string) (*tlsCertificate, bool) {
+	r.mx.Lock()
+	defer r.mx.Unlock()
+
+	cert, ok := r.lookup[key]
+	if !ok || cert == nil {
```

Comment:
Why cert could be nil? Lets ensure we never put nil into the map

Reply:
Now validating that cert is not `nil` before registering.